src/canopy/chat_engine/query_generator/cohere.py

```diff
@@ -0,0 +1,42 @@
+from typing import List, Optional, cast
+
+from canopy.chat_engine.query_generator import QueryGenerator
+from canopy.chat_engine.history_pruner.raising import RaisingHistoryPruner
+from canopy.llm import BaseLLM, CohereLLM
+from canopy.models.data_models import Messages, Query
+
+
+class CohereQueryGenerator(QueryGenerator):
+    """
+    Query generator for LLM clients that have a built-in feature to
+    generate search queries from chat messages.
+    """
+    _DEFAULT_COMPONENTS = {
+        "llm": CohereLLM,
+    }
+
+    def __init__(self,
+                 *,
+                 llm: Optional[BaseLLM] = None):
+        self._llm = llm or self._DEFAULT_COMPONENTS["llm"]()
+
+        if not isinstance(self._llm, CohereLLM):
+            raise NotImplementedError(
+                "CohereQueryGenerator only compatible with CohereLLM"
+            )
+
+        self._history_pruner = RaisingHistoryPruner()
+
+    def generate(self,
+                 messages: Messages,
+                 max_prompt_tokens: int) -> List[Query]:
+        messages = self._history_pruner.build(chat_history=messages,
+                                              max_tokens=max_prompt_tokens)
+        llm = cast(CohereLLM, self._llm)
+        queries = llm.generate_search_queries(messages)
+        return [Query(text=query) for query in queries]
+
+    async def agenerate(self,
+                        messages: Messages,
+                        max_prompt_tokens: int) -> List[Query]:
+        raise NotImplementedError
```

Comment:
same comment as below, but we do have an Async client in the Cohere Python SDK

Reply:
@aalpat4 yes, you are right again. I did notice at some point that there is an async Cohere client. However I also notice that the other LLM backends in Canopy do not yet support the async versions of the functions, and two of the three other query generators do not either, including the default one. I think it would be best to try to get this PR merged in without the async functions first, and do the async implementation in another PR.

Reply:
@scottmx81 @aalpat4 we are planning to implement full Async support across all system parts (embedding, LLM etc) very soon. But as @scottmx81 said, right now it's not a requirement for adding Cohere LLM support.  
We can work on this together as a separate effort later